File: jetbrains/src/main/kotlin/identitybraces/core/Hash.kt
package identitybraces.core

/**
 * Deterministic 64-bit hashing, bit-identical to the Visual Studio extension's
 * `IdentityBraces.Core.Hash` and the VS Code port's `hash.ts`.
 *
 * A brace's identity has to outlive the session, the machine and the editor: the same file
 * opened in Rider, in Visual Studio and in VS Code must hand the same brace the same colour
 * and the same personality. That means the same FNV-1a and the same SplitMix64 finaliser,
 * down to the wrapping arithmetic — and never `String.hashCode`, whose contract says nothing
 * about matching anyone else.
 *
 * Unlike the TypeScript port this is a straight transliteration of the C#: the JVM has real
 * 64-bit integers, and [ULong] makes the shifts and the modulo unsigned without any of the
 * two-halves arithmetic the JavaScript had to do.
 */
internal object Hash {
    private const val FNV_OFFSET_BASIS: ULong = 14695981039346656037uL
    private const val FNV_PRIME: ULong = 1099511628211uL

    /** FNV-1a over the UTF-16 code units of [value]. */
    fun fnv1a(value: CharSequence): ULong {
        var hash = FNV_OFFSET_BASIS
        for (i in 0 until value.length) {
            hash = hash xor value[i].code.toULong()
            hash *= FNV_PRIME
        }

        return hash
    }

    /**
     * SplitMix64 finaliser. Folds [salt] in and avalanches the result, so two hashes that
     * differ in one bit land far apart.
     */
    fun mix(hash: ULong, salt: ULong): ULong {
        var z = hash + salt * 0x9E3779B97F4A7C15uL
        z = (z xor (z shr 30)) * 0xBF58476D1CE4E5B9uL
        z = (z xor (z shr 27)) * 0x94D049BB133111EBuL
        return z xor (z shr 31)
    }

    /** Maps a hash onto [0, 1) using the 53 bits a double can hold exactly. */
    fun toUnitInterval(hash: ULong): Double {
        return (hash shr 11).toDouble() * (1.0 / 9007199254740992.0)
    }

    /** Maps a hash onto [0, [count]). */
    fun toIndex(hash: ULong, count: Int): Int {
        return (hash % count.toULong()).toInt()
    }

    /** The sixteen lower-case hex digits the parity fixture writes an identity as. */
    fun hex(hash: ULong): String {
        return hash.toString(16).padStart(16, '0')
    }
}

File: src/IdentityBraces/Core/Hash.cs
using System.Text;

namespace IdentityBraces.Core
{
    /// <summary>
    /// Deterministic hashing.
    /// <para>
    /// Deliberately not <c>string.GetHashCode</c>: that is randomised per process on
    /// several runtimes, which would hand every brace a brand new personality on each
    /// Visual Studio restart. A brace's identity has to outlive the session.
    /// </para>
    /// </summary>
    internal static class Hash
    {
        private const ulong FnvOffsetBasis = 14695981039346656037UL;
        private const ulong FnvPrime = 1099511628211UL;

        /// <summary>FNV-1a over the UTF-16 code units of <paramref name="value"/>.</summary>
        public static ulong Fnv1a(string value)
        {
            ulong hash = FnvOffsetBasis;
            for (int i = 0; i < value.Length; i++)
            {
                hash ^= value[i];
                hash *= FnvPrime;
            }

            return hash;
        }

        /// <summary>FNV-1a over a <see cref="StringBuilder"/>, avoiding an interim string.</summary>
        public static ulong Fnv1a(StringBuilder value)
        {
            ulong hash = FnvOffsetBasis;
            for (int i = 0; i < value.Length; i++)
            {
                hash ^= value[i];
                hash *= FnvPrime;
            }

            return hash;
        }

        /// <summary>
        /// SplitMix64 finaliser. Folds <paramref name="salt"/> in and avalanches the
        /// result, so two hashes that differ in one bit land far apart.
        /// </summary>
        public static ulong Mix(ulong hash, ulong salt)
        {
            ulong z = hash + salt * 0x9E3779B97F4A7C15UL;
            z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9UL;
            z = (z ^ (z >> 27)) * 0x94D049BB133111EBUL;
            return z ^ (z >> 31);
        }

        /// <summary>Maps a hash onto [0, 1) using the 53 bits a double can hold exactly.</summary>
        public static double ToUnitInterval(ulong hash)
        {
            return (hash >> 11) * (1.0 / 9007199254740992.0);
        }

        /// <summary>Maps a hash onto [0, <paramref name="count"/>).</summary>
        public static int ToIndex(ulong hash, int count)
        {
            return (int)(hash % (ulong)count);
        }
    }
}
